src/primary_backup/main.rs

```diff
@@ -135,130 +114,234 @@ mod tests {
+        assert_get_msg(KEY, VALUE, client_address, false).await;
+    }
 
-        let reply = ClientCommand::Get {
-            key: "k1".to_string(),
-        }
-        .send_to(client_address)
-        .await
-        .unwrap();
-        assert!(reply.is_none());
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_replicated_server() {
+        let (network_address_primary, client_address_primary) = get_address_pair(BASE_PORT + 2);
+        let (network_address_replica, client_address_replica) = get_address_pair(BASE_PORT + 4);
+
+        run_node(
+            db_path("db_test_primary2"),
+            network_address_primary,
+            client_address_primary,
+            None,
+        )
+        .await;
+        run_node(
+            db_path("db_test_backup2"),
+            network_address_replica,
+            client_address_replica,
+            Some(network_address_primary),
+        )
+        .await;
 
-        let reply = ClientCommand::Set {
-            key: "k1".to_string(),
-            value: "v1".to_string(),
-        }
-        .send_to(client_address)
-        .await
-        .unwrap();
+        // get value on primary should be None
+        assert_get_msg(KEY, VALUE, client_address_primary, true).await;
+        // set value on primary
+        assert_set_msg(KEY, VALUE, client_address_primary).await;
+        // get value on primary
+        assert_get_msg(KEY, VALUE, client_address_primary, false).await;
+        // get value on replica to make sure it was replicated
+        assert_get_msg(KEY, VALUE, client_address_replica, false).await;
+    }
 
-        sleep(Duration::from_millis(2000)).await;
-        assert!(reply.is_some());
-        assert_eq!("v1".to_string(), reply.unwrap());
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_backup_change_to_primary() {
+        let (network_address_primary, client_address_primary) = get_address_pair(BASE_PORT + 6);
+        let (network_address_replica, client_address_replica) = get_address_pair(BASE_PORT + 8);
+
+        let (node_handle, _, _) = run_node(
+            db_path("db_test_primary3"),
+            network_address_primary,
+            client_address_primary,
+            None,
+        )
+        .await;
+        run_node(
+            db_path("db_test_backup3"),
+            network_address_replica,
+            client_address_replica,
+            Some(network_address_primary),
+        )
+        .await;
 
-        let reply = ClientCommand::Get {
-            key: "k1".to_string(),
-        }
-        .send_to(client_address)
-        .await
-        .unwrap();
+        tokio::time::sleep(Duration::from_millis(1000)).await;
 
-        sleep(Duration::from_millis(20)).await;
-        assert!(reply.is_some());
-        assert_eq!("v1".to_string(), reply.unwrap());
+        //kill primary
+        node_handle.abort();
+
+        //wait to replica to take the lead
+        tokio::time::sleep(Duration::from_millis(1000 * 9)).await;
```

Comment:
instead of waiting this much (which I presume will make the test flaky when running the test in different environments), can't you do something with retries like this?

https://github.com/Entropy1729/consensus-workbench/blob/91500c174ec53292ee1d312ff0bea9b4c1c2dd94/src/blockchain/main.rs#L268-L287